### ISSUE TITLE
Different transition behavior between {{action}} and {{fn}}

### ISSUE THREAD

**Issue by maxwondercorn** [emberjs/ember.js#18801]
Ember 3.12.2

While converting from `{{action}}` to the `{{fn}}` helper I have several routes that now abort during the transition. The behavior is inconsistent and is affecting three routes out of a dozen similar routes. 

The three routes will only transition the first time after the app is refreshed. After the first transition they will fail.
 
The route transitions are in controller actions.

```js
  // controller
  @action
  addNodel() {
    this.transitionToRoute('setup.nodes.new');
  }
```
Execution using `{{action}}`

Edit: removed `onclick=` (see edit below)
```js
<a href="#" {{action @add bubbles=false}}>
  Add {{@name}}
</a>
```

Converted to `{{fn}}`
```js
<a href="#" {{on 'click' (fn @add) bubbles=false}}>
  Add {{@name}}
</a>
```
**{{action}} Transition log - all routes work**

First transition - setup.nodes.new
```sh
Attempting transition to setup.nodes.new
Preparing to transition from 'setup.nodes.index' to 'setup.nodes.new'
Transition #1: Resolved all models on destination route; finalizing transition.
Transitioned into 'setup.nodes.new'
Transition #1: TRANSITION COMPLETE.
```
Subsequent transitions
```sh
Attempting transition to setup.nodes.new
Preparing to transition from 'setup.nodes.index' to 'setup.nodes.new'
Attempting URL transition to /setup/nodes?site=ALL
Transition #3: setup.nodes.new: transition was aborted
Transition #3: detected abort.
Transition #4: Resolved all models on destination route; finalizing transition.
Transitioned into 'setup.nodes.index'
Transition #4: TRANSITION COMPLETE.
```

**{{fn}} Transition log for failing routes - setup.nodes.new**

First transition
```sh
Attempting transition to setup.nodes.new
Preparing to transition from 'setup.nodes.index' to 'setup.nodes.new'
Transition #1: Resolved all models on destination route; finalizing transition.
Transitioned into 'setup.nodes.new'
Transition #1: TRANSITION COMPLETE.
```
Subsequent transitions
```sh
Attempting transition to setup.nodes.new
Preparing to transition from 'setup.nodes.index' to 'setup.nodes.new'
Attempting URL transition to /setup/nodes?site=ALL
Transition #3: setup.nodes.new: transition was aborted
Transition #4: Resolved all models on destination route; finalizing transition.
Transitioned into 'setup.nodes.index'
Transition #4: TRANSITION COMPLETE.
Transition #3: detected abort.
```

Reviewing a working {{fn}} route's log (below) the URL transition is attempted at the end of the tranistion. For all practical purposes the routes/logic are the same so there is nothing obvious.  Since transitions are promises do they reject on an abort?  i.e. would a `.catch()` provide additional information?

Edit: I corrected the working example the `<a>` tag uses `{{action ...}}`, not `onclick={{action...}}`.  `onclick=` behaves the same way as `{{on}}` so the issue seems to be related to using a closure action. 

First transition - setup.materials.new
```sh
Attempting transition to setup.materials.new
Preparing to transition from 'setup.materials.index' to 'setup.materials.new'
Transition #1: Resolved all models on destination route; finalizing transition.
Transitioned into 'setup.materials.new'
Transition #1: TRANSITION COMPLETE.
```
Subsequent transitions
```sh
Attempting transition to setup.materials.new
Preparing to transition from 'setup.materials.index' to 'setup.materials.new'
Transition #3: Resolved all models on destination route; finalizing transition.
Transitioned into 'setup.materials.new'
Transition #3: TRANSITION COMPLETE.
Attempting URL transition to /setup/materials/new?site=ALL
```

**Comment by jelhan** [comment]
The `{{on}}` modifier does not have a `bubbles` argument as far as I'm aware. I think you need to call `event.stopPropagation():` instead.

**Comment by jherdman** [comment]
Alternatively https://github.com/buschtoens/ember-event-helpers

**Comment by maxwondercorn** [comment]
@jelhan I'm not sure I need the `bubbles` anymore. I believe the reason it's there has been refactored away. Hope to find time in the next couple weeks to dig back into the issue

@jherdman Thanks for info on `ember-event-helpers`.  I didn't know they existed.

**Comment by locks** [comment]
Hm, there's no reason to use `fn` if you are not passing arguments. According to your code samples, you are not calling `event.preventDefault()` either, so I'm surprised it works at all 🤔 
Can you put together a small reproduction?

**Comment by maxwondercorn** [comment]
@locks @jelhan @jherdman Thanks all for the input on this issue.  

We've been working on a backend refactor so this was shelved for awhile. We haven't been able create a simple reproduction but we have determined the issue has to do with using RSVP in the failing routes.

Basically the model hook fails aborting the transaction. We did do some tracing through the run-loop, RSVP, etc but that's not a lot of "fun".

In the end, we converted the offending from action on an <a> tag to a <LinkTo>. This works and is more idiomatic Ember.  Another advantage is that we were able to eliminate number of controllers with some other minor refactoring.

I'm closing this issue
